contrib/babelfishpg_tsql/src/pltsql_bulkcopy.c

```diff
@@ -703,6 +703,15 @@ ExecuteBulkCopy(BulkCopyState cstate, int rowCount, int colCount,
 					else
 					{
 						myslot->tts_values[col_index_to_insert] = Values[col_index_to_fetch];
+
+						/* In case of identity column, store the updated identity sequence value. */
+						if (cstate->seq_index == i)
```

Comment:
What will happen if in a concurrent session we do a simple insert in this table, how can we handle such scenarios? what is TSQL behaviour?

Reply:
In the scenario, where performing a bulk insert from a source table to a destination table, and there is a concurrent insert happening in the source table, the impact on the identity values in the destination table will depend on the order of operations.

Here's what could happen:

1. `Bulk Insert Happens First`: If the bulk insert operation from the source table to the destination table completes before the concurrent insert in the source table, the identity values in the destination table will be consistent with the source table at the time of the bulk insert.
In this case, the concurrent insert in the source table will not affect the identity values in the destination table because the bulk insert has already copied the data, including the identity values, from the source table.

2. `Concurrent Insert Happens First`: If the concurrent insert in the source table happens before the bulk insert operation, the identity values in the destination table may be affected.
When the bulk insert operation copies data from the source table, it will include the newly inserted row with the new identity value from the concurrent insert. As a result, the identity value sequence in the destination table will have a gap or a duplicate value, depending on how the bulk insert process handles identity values.

Here's what we have done to handle these scenarios:

1. For Scenario-1, if Bulk insert happens first then obviously the flow will continue and there will be no implications.

2. However for Scenario-2, if Insert happens before seeding the identity value, then we will consider the inserted row to calculate the new_identity_value during bulk insert and then while seeding the identity value we will impose lock on the table to prevent further inserts until the end of bulk insert process.

Reply:
> If the concurrent insert in the source table happens

The question was about concurrent insert in destination table. What will happen if I do concurrent BCP on same destination table? Our logic today for identity does not take care of this.

Reply:
In case of concurrent insert in destination table:
As per our current implementation, we seed the identity sequence value at the end of BCP execution.
So, having concurrent insert will insert the row with identity sequence value incremented from last seed value and BCP will insert rows with given identity values (keep identity is ON), so there might be difference in Identity sequence value as compared to TSQL which will hold Row lock in each row inserted by BCP, seeding the new value each time, so the concurrent insert will have identity sequence value incremented from new seed value.
Consider this scenario:
Session 1 executes:
```
SET IDENTITY_INSERT MyTable ON;
INSERT INTO MyTable (Id, Name) VALUES (10, 'A'), (11, 'B');
SET IDENTITY_INSERT MyTable OFF;
```
Session 2 executes concurrently:
```
INSERT INTO MyTable (Name) VALUES ('C');
```
Outcome: (TSQL)

- Session 1 inserts records with explicit identity values (10, 11).
- Session 2 inserts a record without specifying an identity value. TSQL will generate the next value in the sequence for the identity column (e.g., if the next value should be 12, then 12 will be used).

whereas in BBF for session since we haven't yet seeded the new value from session-1 i.e. 11, so session will insert value with old seed value(e.g., if the next value should be 1, then 1 will be used).

To resolve this we can have seed value updated on each row copied, but it will have it's performance overhead.

Reply:
Need to update this thread about our discussion. We will keep working on this and let this PR get merged for now